cpp/src/arrow/memory_pool.h

```diff
@@ -63,6 +63,42 @@ class MemoryPoolStats {
 /// take care of the required 64-byte alignment.
 class ARROW_EXPORT MemoryPool {
  public:
+  class ARROW_EXPORT ImmutableZeros {
```

Comment:
I'm probably missing something, but why is the existing Buffer interface not sufficient for this?

Reply:
I think you're right, actually. My reasoning for introducing it was that I needed something to model shared ownership of a region of zero-allocated (virtual) memory for a bunch of Buffers (`ImmutableZerosPoolBuffer` specifically, as created using `MakeBufferOfZeros()`), so some RAII object inside a `shared_ptr` would do the trick. I didn't even consider that that RAII object could also just be a Buffer implementation, with its implementation details hidden that way. It felt very wrong to define a public class for something this specific in such a prominent place to begin with, so I'm glad to get rid of it... The current `ImmutableZerosPoolBuffer` class could then be used much more generically for shared-ownership views of some buffer, unless that already exists (?), in which case I can just reuse that.

I don't think I'll get around to refactoring this today, but I'll take a closer look tomorrow (and then rebase after that).

Reply:
Okay, `ImmutableZeros` now implements `Buffer` and has its implementation hidden.

As I looked at the Buffer interface a bit better I also figured a specialized memory manager was probably in order, since these buffers have different, managed allocate operations. With that, I *think* the interface for using immutable zero CPU buffers is now fully symmetric to the one for regular, mutable CPU buffers.

81aa078